### ISSUE TITLE
pyright does not correctly work with pygit2 if empty pyrightconfig.json is present in the folder

### ISSUE THREAD

**Issue by tupos** [microsoft/pyright#9104]
### Version
```
2024-09-28 14:07:31.067 [info] (Client) Pylance async client (2024.9.2) started with python extension (2024.14.1)
2024-09-28 14:07:31.124 [info] [Info  - 14:07:31] (86171) Pylance language server 2024.9.2 (pyright version 1.1.378, commit ce64adc1) starting
```

python is from homebrew of mac os 14.6.7

### Steps to reproduce
```bash
mkdir 111
cd 111
python3 -m venv .venv
. ./.venv/bin/activate
pip install pygit2
echo $'import pygit2\n\nrepo: pygit2.Repository = pygit2.init_repository("/path")\n' > tt.py
echo $'{\n}' > pyrightconfig.json
```
Open folder `111` in vscode.

After this in the file `tt.py` we get an error
```
Type "pygit2.repository.Repository | pygit2._pygit2.Repository" is not assignable to declared type "Repository"
  Type "pygit2.repository.Repository | pygit2._pygit2.Repository" is not assignable to type "Repository"
    "pygit2._pygit2.Repository" is not assignable to "pygit2.repository.Repository"Pylance[reportAssignmentType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportAssignmentType)
```

However the error is wrong as the function returns the `Repository`. Also if we go to definition of `init_repository` we can see some undefined symbols.

If `pyrightconfig.json` is removed the completion works as expected.

### What to expect

If `pyrightconfig.json` is present in the root of the folder pyright should not report wrong error.

**Comment by erictraut** [comment]
Pyright is working as intended here. This appears to be a bug in the typeshed stubs for pygit2.

The pygit2 library has type stubs that are included in the typeshed repository. Pyright bundles all third-party typeshed stubs. 

The `__init__.pyi` stub in this case is defining the symbol `Repository` twice. The first time is [here](https://github.com/python/typeshed/blob/119cd09655dcb4ed7fb2021654ba809b8d88846f/stubs/pygit2/pygit2/__init__.pyi#L8), and the second time is [here](https://github.com/python/typeshed/blob/119cd09655dcb4ed7fb2021654ba809b8d88846f/stubs/pygit2/pygit2/__init__.pyi#L30). These two definitions provide different types, so pyright is combining them into a union. 

It appears that this hack was added to work around a bug in mypy, but this isn't a valid workaround as it relies on (buggy) behavior that is specific to mypy. If you comment out the second definition of `Repository` in the stub file, it works fine with pyright. I think this should work fine with mypy as well, but I haven't confirmed.

I recommend that you report the issue to the maintainers of typeshed and/or file a PR in that project to fix this issue.

Pyright regularly syncs with typeshed, so any fix in typeshed will show up within a week or two within pyright.

**Comment by tupos** [comment]
Hello. Thanks for a quick reply.

But why is it working correctly if there is no `pyrightconfig.json` present in the root folder of the project?

Also why if the `pyrightconfig.json` is present, there are undefined symbols in `__init__.py` of pygit2 (for example enums.Feature `"Feature" is not a known attribute of module "pygit2"Pylance[reportAttributeAccessIssue](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportAttributeAccessIssue)` ? 

Is it also expected?

**Comment by erictraut** [comment]
> Is it also expected?

Yes, this is expected behavior. If you add a `pyrightconfig.json` to your project, you will get pyright's default type checking behaviors in your project. This means the "typeCheckingMode" defaults to "standard", which enables a bunch of checks.

If you do not provide a `pyrightconfig.json`, then you can specify type checking behaviors in your LSP settings. For example, you can set "typeCheckingMode" to "off", and you won't see any typing-related errors. If you have a `pyrightconfig.json` present, it overrides type checking settings in your LSP settings. The idea is that if a config file is present, you will want all collaborators on the team to see the same type checking behaviors in their editors — and this should also be consistent with their CI scripts. If you'd like to disable all type checking diagnostics for your project, you can add `"typeCheckingMode": "off"` to your `pyrightconfig.json` file.
